saleor/discount/utils.py

```diff
@@ -369,16 +377,21 @@ def create_discount_objects_for_catalogue_promotions(
         discount_amount = _get_discount_amount(line_info.channel_listing, line.quantity)
 
         # get the existing discounts for the line
-        discounts_to_update = line_info.get_promotion_discounts()
+        discounts_to_update = line_info.get_catalogue_discounts()
         rule_id_to_discount = {
             discount.promotion_rule_id: discount for discount in discounts_to_update
         }
 
-        # delete all existing discounts if the line is not discounted
-        if not discount_amount:
+        # delete all existing discounts if the line is not discounted or it is a gift
+        if not discount_amount or line.is_gift:
             ids_to_remove = [discount.id for discount in discounts_to_update]
-            CheckoutLineDiscount.objects.filter(id__in=ids_to_remove).delete()
-            line_info.discounts = []
+            if ids_to_remove:
+                CheckoutLineDiscount.objects.filter(id__in=ids_to_remove).delete()
```

Comment:
Is it expected that you're calling the db query in the loop, line by line?

Reply:
fixed